### PR TITLE
Sheet: internal refactors

### PR DESCRIPTION
**Pull request by AlbertCarreras**
#### What changed?

Sheet: internal refactors

**1st commit**
- Split code in Sheet & Internal Sheet
- Implement InternalDismissButton
- Fix Flow type (which required adding error messaging for `subHeading` w/o `heading` props) to allow autogenerated props, implemented autogenerated props
- Implemented Sheet's `accessibilityDismissButtonLabel` into DefaultLabelProvider
- Fix tests
- Cleanup AnimationController to use useAnimationController rather than render props pattern to access `onDismissStart`

**2nd commit**
- Removed unnecessary render props, keeping basic provider wrapping InternalSheet

**3rd commit**
- Overall renaming for better comprehension of animation flow

**4th commit**
- Reduce animation states to 4: `opening`, `closing`, `null`, and `unmount`

Making it easier to read and follow logic so we an update the animation logic of the component.

#### Why?

Prepare code for more clarity and future work. Nice cleanup

For testing this changes toggle reduce motion in your laptop
Screenshot 2023-01-24 at 6 51 25 PM

### CODE REVIEW

File: docs/examples/sheet/defaultExample.js [hunk 1]
```diff
@@ -180,7 +27,160 @@ export default function AccessibilityExample(): Node {
       </Box>
       {shouldShow && (
         <Layer zIndex={sheetZIndex}>
-          <SheetWithAccessibilityLabels onDismiss={() => setShouldShow(false)} />
+          <Sheet
+            accessibilityDismissButtonLabel="Close audience creation sheet"
+            accessibilitySheetLabel="Audience list creation for new campaign"
+            heading="Create a new audience list"
+            onDismiss={() => {}}
+            // eslint-disable-next-line react/no-unstable-nested-components
```
Comment (dangerismycat) [inline docs/examples/sheet/defaultExample.js:35]: This can be solved by adding `useCallback`

File: docs/pages/web/sheet.js [hunk 2]
```diff
@@ -21,7 +21,182 @@ export default function SheetPage({ generatedDocGen }: {| generatedDocGen: DocGe
 
   return (
     <Page title={generatedDocGen?.displayName}>
-      <PageHeader name={generatedDocGen?.displayName} description={generatedDocGen?.description}>
+      <PageHeader
+        name={generatedDocGen?.displayName}
+        description={generatedDocGen?.description}
+        defaultCode={`\
```
Comment (dangerismycat) [inline docs/pages/web/sheet.js:27]: typo: stray `\` character

File: packages/gestalt/src/Sheet.js [hunk 3]
```diff
@@ -1,321 +1,90 @@
...
-
-type OnAnimationEndStateType = 'in' | 'out';
+type NodeOrRenderProp = Node | (({| onDismissStart: () => void |}) => Node);
 
-type SheetMainProps = {|
+type Props = {|
+  /**
+    Supply a short, descriptive label for screen-readers as a text alternative to the Dismiss button. See the [Accessibility section](#Accessibility) for more info.
+   */
   accessibilityDismissButtonLabel: string,
```
Comment (dangerismycat) [inline packages/gestalt/src/Sheet.js:13]: This should now be optional, yes?

File: packages/gestalt/src/Sheet.js [hunk 4]
```diff
@@ -1,321 +1,90 @@
...
-    heading = undefined,
-    size,
-    subHeading = undefined,
-  } = props;
-
+function Sheet({
+  accessibilityDismissButtonLabel,
+  accessibilitySheetLabel,
+  children,
+  closeOnOutsideClick = true,
```
Comment (dangerismycat) [inline packages/gestalt/src/Sheet.js:62]: We should really update this prop name at some point — boolean props that default to `true` is a bit of an anti-pattern. This should really be something like `disableCloseOnOutsideClick`. (Totally out of scope for this PR!)

**Review by dangerismycat — APPROVED** [review]

File: docs/examples/sheet/defaultExample.js (hunk 1 above)
Comment (AlbertCarreras) [reply docs/examples/sheet/defaultExample.js:35]: done

File: docs/pages/web/sheet.js (hunk 2 above)
Comment (AlbertCarreras) [reply docs/pages/web/sheet.js:27]: DONE

File: packages/gestalt/src/Sheet.js (hunk 3 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/Sheet.js:13]: DONE, good catch

**Pull request merged**